### PR TITLE
Replace `HashMap` with `IndexMap` in pattern binding resolve 

### PR DESCRIPTION
**Pull request by Noratrieb**
fixes https://github.com/rust-lang/rust/pull/114332#discussion_r1284189179

### CODE REVIEW

File: compiler/rustc_resolve/src/late.rs [hunk 1]
```diff
@@ -3207,7 +3207,7 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
 
         // 3) Report all missing variables we found.
         let mut missing_vars = missing_vars.into_iter().collect::<Vec<_>>();
-        missing_vars.sort_by_key(|&(sym, ref _err)| sym);
+        missing_vars.sort_by(|(sym_a, _), (sym_b, _)| sym_a.as_str().cmp(sym_b.as_str()));
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late.rs:3210]: Should `missing_vars` be a `FxIndexMap` from the beginning, to have them in source order and avoid any sorting?

Comment (cjgillot) [reply compiler/rustc_resolve/src/late.rs:3210]: Likewise for `inconsistent_vars`?

**Review by compiler-errors — COMMENTED** [review]
r=me after cjgillot's nit, i agree that's probably the better approach

File: compiler/rustc_resolve/src/late.rs (hunk 1 above)
Comment (Nilstrieb) [reply compiler/rustc_resolve/src/late.rs]: makes sense

File: compiler/rustc_resolve/src/late.rs [hunk 2]
```diff
@@ -3212,8 +3212,7 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
         }
 
         // 3) Report all missing variables we found.
-        let mut missing_vars = missing_vars.into_iter().collect::<Vec<_>>();
-        missing_vars.sort_by_key(|&(sym, ref _err)| sym);
+        let missing_vars = missing_vars.into_iter().collect::<Vec<_>>();
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late.rs:3215]: We can directly iterate on the indexmap, without collecting beforehand.

File: compiler/rustc_resolve/src/late.rs [hunk 3]
```diff
@@ -3226,8 +3225,7 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
         }
 
         // 4) Report all inconsistencies in binding modes we found.
-        let mut inconsistent_vars = inconsistent_vars.iter().collect::<Vec<_>>();
-        inconsistent_vars.sort();
+        let inconsistent_vars = inconsistent_vars.iter().collect::<Vec<_>>();
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late.rs:3228]: Likewise.

**Comment by Nilstrieb** [comment]
i get different results on host x86-64 targetting host and targetting wasm32... what????

**Comment by Nilstrieb** [comment]
there was another hashmap hiding in `binding_mode_map`. Don't do hashmaps, kids.

**Comment by cjgillot** [comment]
A `#[deny(rustc::potential_query_instability)]` should help to find those.

**Pull request merged**